Custom element upgrades

cbrewster opened the pull request:
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17935)
<!-- Reviewable:end -->

cbrewster committed:
Add construction stack

cbrewster committed:
Implement element upgrade algorithm

cbrewster committed:
Add upgrade reaction

cbrewster committed:
Enqueue upgrades

cbrewster committed:
Use HTMLElement interface for possible CEs

cbrewster committed:
Fix compartment mismatch issue

cbrewster committed:
Check namespace during ce def lookup

cbrewster committed:
Handle exceptions during upgrades

cbrewster committed:
Track custom element state

highfive commented:
Heads up! This PR modifies the following files:
 * @fitzgen: components/script/dom/customelementregistry.rs, components/script/dom/document.rs, components/script/dom/element.rs, components/script/dom/bindings/codegen/CodegenRust.py, components/script/dom/create.rs and 4 more
 * @kichjang: components/script/dom/customelementregistry.rs, components/script/dom/document.rs, components/script/dom/element.rs, components/script/dom/bindings/codegen/CodegenRust.py, components/script/dom/create.rs and 4 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

cbrewster commented:
r? @jdm

jdm commented on components/script/dom/bindings/interface.rs:
```diff
@@ -281,24 +282,42 @@ pub unsafe fn html_constructor<T>(window: &Window, call_args: &CallArgs) -> Fall
+            Root::downcast(element).ok_or(Error::InvalidState)
+        },
+        // Step 10
+        Some(ConstructionStackEntry::AlreadyConstructedMarker) => return Err(Error::InvalidState),
```
No need for `return` here.

cbrewster replied:
Done.

jdm commented on components/script/dom/customelementregistry.rs:
```diff
@@ -462,6 +462,72 @@ impl CustomElementDefinition {
+            // TODO: Catch exceptions
+            return;
+        }
+        // Step 8
```
This is 7.2.

cbrewster replied:
Done.

jdm commented on components/script/dom/customelementregistry.rs:
```diff
@@ -462,6 +462,72 @@ impl CustomElementDefinition {
+        // Go into the constructor's compartment
+        let _ac = JSAutoCompartment::new(cx, definition.constructor.callback());
+        let args = HandleValueArray::new();
+        // Step 7
```
7.1

cbrewster replied:
Done.

jdm commented on components/script/dom/customelementregistry.rs:
```diff
@@ -462,6 +462,72 @@ impl CustomElementDefinition {
+            return;
+        }
+        if !same {
+            // TODO: Throw InvalidStateError
```
?

cbrewster replied:
Sorry, this was implemented in a later commit.

jdm commented on components/script/dom/customelementregistry.rs:
```diff
@@ -79,13 +79,19 @@ impl CustomElementRegistry {
+                             local_name: &LocalName,
+                             is: Option<&LocalName>)
                              -> Option<Rc<CustomElementDefinition>> {
+        // Step 1
```
Let's move this to Document::lookup_custom_element_definition instead.

cbrewster replied:
Done.

jdm commented on components/script/dom/customelementregistry.rs:
```diff
@@ -287,17 +288,28 @@ impl CustomElementRegistryMethods for CustomElementRegistry {
+            let is = candidate.get_is();
+            if *candidate.local_name() == local_name &&
+                *candidate.namespace() == ns!(html) &&
+                (is.is_none() || is.as_ref() == Some(&name))
```
Don't we need to check `extends` here?

cbrewster replied:
Oops, yup. Fixed.

jdm commented on components/script/dom/customelementregistry.rs:
```diff
@@ -492,48 +492,70 @@ pub fn upgrade_element(definition: Rc<CustomElementDefinition>, element: &Elemen
+}
+
+#[allow(unsafe_code)]
+fn run_upgrade_constructor(constructor: &Rc<Function>, element: &Element) -> ErrorResult {
```
Add a spec link here.

cbrewster replied:
Done.

jdm commented on components/script/dom/element.rs:
```diff
@@ -150,6 +150,7 @@ pub struct Element {
     /// https://dom.spec.whatwg.org/#concept-element-custom-element-definition
     #[ignore_heap_size_of = "Rc"]
     custom_element_definition: DOMRefCell<Option<Rc<CustomElementDefinition>>>,
+    custom_element_state: Cell<CustomElementState>,
```
Add a link to https://dom.spec.whatwg.org/#concept-element-custom-element-state .

cbrewster replied:
Done.

jdm commented on components/script/dom/element.rs:
```diff
@@ -179,6 +180,14 @@ pub enum CustomElementCreationMode {
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, HeapSizeOf, JSTraceable)]
+pub enum CustomElementState {
```
Add a link to https://dom.spec.whatwg.org/#concept-element-custom-element-state .

cbrewster replied:
Done.

jdm reviewed (changes requested):
Great work! These changes were quite readable and straightforward to review.

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 0c2725d has been approved by `jdm`

<!-- @bors-servo r=jdm 0c2725d6269a10552ef01b0d18c731fd290edeb4 -->

bors-servo commented:
:hourglass: Testing commit 0c2725d6269a10552ef01b0d18c731fd290edeb4 with merge 866c8e958e05aeb74d6732a8496a34b6f50d887e...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5392)

jdm commented:
```
  ▶ Unexpected subtest result in /_mozilla/mozilla/collections.html:
  │ FAIL [expected PASS] test_binding\n    page   43
  │   → assert_true: Should be function HTMLUnknownElement() {
    [native code]
} expected true got false
  │ 
  │ check_collection/<@http://web-platform.test:8000/_mozilla/mozilla/:86:11
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ check_collection@http://web-platform.test:8000/_mozilla/mozilla/:71:5
  │ check_tag@http://web-platform.test:8000/_mozilla/mozilla/:95:7
  │ @http://web-platform.test:8000/_mozilla/mozilla/:145:3
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  └ @http://web-platform.test:8000/_mozilla/mozilla/:69:1

  ▶ Unexpected subtest result in /_mozilla/mozilla/prototypes.html:
  │ FAIL [expected PASS] Untitled
  │   → assert_true: Should be HTMLUnknownElement expected true got false
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/prototypes.html:21:3
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  └ @http://web-platform.test:8000/_mozilla/mozilla/prototypes.html:10:1
```

cbrewster committed:
Change foo-á to fooá

These tests assumed that foo-á would create a HTMLUnknownElement; however,
foo-á is a valid custom element name; therefore, according to the spec, a
HTMLElement instace is created instead. I changed the name to fooá so a
HTMLUnknownElement is created as the test expects.

cbrewster commented:
@bors-servo r=jdm

bors-servo commented:
:pushpin: Commit aa9b9c1 has been approved by `jdm`

<!-- @bors-servo r=jdm aa9b9c10f315179df5d9fd6dcbf8a36129092b19 -->

bors-servo commented:
:hourglass: Testing commit aa9b9c10f315179df5d9fd6dcbf8a36129092b19 with merge 931e205033bb5a68a1209af020fa73267db5f53a...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5360)

jdm commented:
@bors-servo: retry

bors-servo commented:
:hourglass: Testing commit aa9b9c10f315179df5d9fd6dcbf8a36129092b19 with merge a087fee4b4daa76c7e1677b0adae512fb6c87d04...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/8256), [arm32](http://build.servo.org/builders/arm32/builds/8480), [arm64](http://build.servo.org/builders/arm64/builds/8513), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8598), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/5362), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/5396), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8596), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/1072), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/1075), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/5296), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/5354), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/1074), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/1078), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3838)
Approved by: jdm
Pushing a087fee4b4daa76c7e1677b0adae512fb6c87d04 to master...
